### PR TITLE
add first vertion of rpc

### PR DESCRIPTION
**Pull request by libenchao**
## summary:
1. add `RPC`, `DataNodeProtocol`, `Server`, `Client`, `Param`
2. modify `NameNode`, `DataNode` to test rpc

## problems:
1. haven't considered concurrency.
2. haven't considered primitive types and `null`

## reviewers
- [ ] @yanyiping 
- [ ] @songpengwei

### CODE REVIEW

**Comment by songpengwei** [comment]
Perfect Code!
I have somequestions.
1. Why not implements Runnable instead of extends Thread? The former provides more expansibility.
2. I think the DataNodeProtocol Class can be renamed. Because it means whose method they(client and server) will call instead of only one peer. Maybe MessageProtocol will be better?  
3. Then the DataNode will not implements the DataNodeProtocol but use one of its implementation classes as a field. I think field is better than inherit and the semantic is more clear.

**Comment by songpengwei** [comment]
Perfect Code!
I have some questions.
1.  Why not implements Runnable instead of extending the class Thread? The former provides more expansibility.
2. I think the DataNodeProtocol Class can be renamed. Because it means whose method they(client and server) will call instead of only one peer. Maybe MessageProtocol will be better?
3. Then the DataNode will not implement the DataNodeProtocol but use one of its implementation classes as a field. I think field is better than inheritment and the semantic is more clear.

File: src/main/java/org/tridiots/ipc/Client.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        Param param = new Param(methodName, paramTypes, params);
+        try {
+            oos.writeObject(param);
+            return ois.readObject();
+        } catch (IOException e) {
+            e.printStackTrace();
+        } catch (ClassNotFoundException e) {
+            e.printStackTrace();
+        }
+        return null;
```
Comment (libenchao) [inline src/main/java/org/tridiots/ipc/Client.java:39]: in "files changed" tab, you can see all the codes changed. When you put your cursor on the changes, you can see:
image

Comment (libenchao) [reply src/main/java/org/tridiots/ipc/Client.java:39]: @songpengwei

**Comment by libenchao** [comment]
@songpengwei thanks for the review.
1. it's a very good idea. Actually, this is way that the `HDFS` uses.
2. I think `DataNodeProtocol` is a good name, because any class implements this protocol can be remotely called by `DataNode`. What's more, we will have `ClientProtocol` for the `Client`'s remote call later.
3. 'field is better than inheritance' is right. but here the `DataNodeProtocol` will use all the information which `NameNode` has, `field` is a little inconvenient.

PS: you can write `markdown` in github, it's wonderful. 

@yanyiping you can review these codes and ask anything or suggest anything. We'll be very happy to hear from you. This is the way we **make progress**.

**Comment by libenchao** [comment]
@songpengwei @yanyiping codes updated. Thanks for the review.

**Pull request merged**